File: facebook/src/test/java/com/facebook/appevents/codeless/ViewHierarchyTest.java
/*
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * You are hereby granted a non-exclusive, worldwide, royalty-free license to use,
 * copy, modify, and distribute this software in source code or binary form for use
 * in connection with the web services and APIs provided by Facebook.
 *
 * As with any software that integrates with the Facebook platform, your use of
 * this software is subject to the Facebook Developer Principles and Policies
 * [http://developers.facebook.com/policy/]. This copyright notice shall be
 * included in all copies or substantial portions of the software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package com.facebook.appevents.codeless;

import static org.junit.Assert.assertTrue;

import android.content.Context;
import android.view.View;
import android.view.ViewParent;
import android.widget.AdapterView;
import androidx.core.view.NestedScrollingChild;
import com.facebook.appevents.codeless.internal.ViewHierarchy;
import com.facebook.internal.Utility;
import java.lang.reflect.Method;
import org.json.JSONObject;
import org.junit.Test;
import org.mockito.Mock;
import org.powermock.api.mockito.PowerMockito;
import org.powermock.core.classloader.annotations.PrepareForTest;

@PrepareForTest({
  ViewHierarchy.class,
})
public class ViewHierarchyTest extends CodelessTestBase {

  @Test
  public void testGetDictionaryOfView() throws Exception {
    JSONObject dict = ViewHierarchy.getDictionaryOfView(root);

    String outerText = dict.getJSONArray("childviews").getJSONObject(0).getString("text");
    assertTrue(outerText.equalsIgnoreCase(Utility.sha256hash("Outer Label")));

    String innerText =
        dict.getJSONArray("childviews")
            .getJSONObject(1)
            .getJSONArray("childviews")
            .getJSONObject(0)
            .getString("text");
    assertTrue(innerText.equalsIgnoreCase(Utility.sha256hash("Inner Label")));
  }

  abstract static class TestAdapterView extends AdapterView implements ViewParent {
    public TestAdapterView(Context context) {
      super(context);
    }
  }

  abstract static class TestNestedScrollingChild implements ViewParent, NestedScrollingChild {}

  @Mock View mockView;
  @Mock TestNestedScrollingChild mockTestNestedScrollingChild;
  @Mock TestAdapterView mockTestAdapterView;
  @Mock ViewParent mockViewParent;

  @Test
  public void testIsAdapterViewItem() throws Exception {
    PowerMockito.spy(ViewHierarchy.class);
    Method isAdapterViewItem =
        ViewHierarchy.class.getDeclaredMethod("isAdapterViewItem", View.class);
    isAdapterViewItem.setAccessible(true);

    // mock NestedScrollingChild -> true
    PowerMockito.when(mockView.getParent()).thenReturn(mockTestNestedScrollingChild);
    assertTrue((boolean) isAdapterViewItem.invoke(ViewHierarchy.class, mockView));

    // mock AdapterView -> true
    PowerMockito.when(mockView.getParent()).thenReturn(mockTestAdapterView);
    assertTrue((boolean) isAdapterViewItem.invoke(ViewHierarchy.class, mockView));

    // mock other cases -> false
    PowerMockito.when(mockView.getParent()).thenReturn(mockViewParent);
    assertTrue(!(boolean) isAdapterViewItem.invoke(ViewHierarchy.class, mockView));
  }
}

File: facebook-core/src/test/kotlin/com/facebook/appevents/codeless/internal/ViewHierarchyTest.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.appevents.codeless.internal

import android.content.Context
import android.view.View
import android.view.ViewParent
import android.widget.Adapter
import android.widget.AdapterView
import androidx.core.view.NestedScrollingChild
import com.facebook.appevents.codeless.CodelessTestBase
import com.facebook.appevents.codeless.internal.ViewHierarchy.getDictionaryOfView
import com.facebook.internal.Utility.sha256hash
import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Test
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever
import org.powermock.api.mockito.PowerMockito

class ViewHierarchyTest : CodelessTestBase() {
  private lateinit var mockView: View
  private lateinit var mockTestNestedScrollingChild: TestNestedScrollingChild
  private lateinit var mockTestAdapterView: TestAdapterView
  private lateinit var mockViewParent: ViewParent

  @Before
  override fun setup() {
    super.setup()
    mockView = mock()
    mockTestNestedScrollingChild = mock()
    mockTestAdapterView = mock()
    mockViewParent = mock()
  }

  @Test
  fun testGetDictionaryOfView() {
    val dict = getDictionaryOfView(root)
    val outerText = dict.getJSONArray("childviews").getJSONObject(0).getString("text")
    assertThat(outerText.equals(sha256hash("Outer Label"), ignoreCase = true)).isTrue
    val innerText =
        dict
            .getJSONArray("childviews")
            .getJSONObject(1)
            .getJSONArray("childviews")
            .getJSONObject(0)
            .getString("text")
    assertThat(innerText.equals(sha256hash("Inner Label"), ignoreCase = true)).isTrue
  }

  abstract class TestAdapterView(context: Context?) : AdapterView<Adapter>(context), ViewParent
  abstract class TestNestedScrollingChild : ViewParent, NestedScrollingChild

  @Test
  fun testIsAdapterViewItem() {
    PowerMockito.spy(ViewHierarchy::class.java)
    val isAdapterViewItem =
        ViewHierarchy::class.java.getDeclaredMethod("isAdapterViewItem", View::class.java)
    isAdapterViewItem.isAccessible = true

    // mock NestedScrollingChild -> true
    whenever(mockView.parent).thenReturn(mockTestNestedScrollingChild)
    assertThat(isAdapterViewItem.invoke(ViewHierarchy::class.java, mockView) as Boolean).isTrue

    // mock AdapterView -> true
    whenever(mockView.parent).thenReturn(mockTestAdapterView)
    assertThat(isAdapterViewItem.invoke(ViewHierarchy::class.java, mockView) as Boolean).isTrue

    // mock other cases -> false
    whenever(mockView.parent).thenReturn(mockViewParent)
    assertThat(isAdapterViewItem.invoke(ViewHierarchy::class.java, mockView) as Boolean).isFalse
  }
}
